docs/DeveloperGuide.adoc

```diff
@@ -767,16 +821,255 @@ Use case ends.
+2.   AddressBook displays the details of the person with the link.
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 1a. The list is empty.
++
+Use case ends.
+
+* 1b. The given name(s) is invalid
++
+Use case ends.
+
+[discrete]
+=== Use case: Sort person by name
+
+*MSS*
+
+1.  User requests to sort person by name
+2.  AddressBook sorts all the contacts
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The list is empty.
++
+Use case ends.
+
+[discrete]
+=== Use case: Sort person by fields in alphabetical order
+
+*MSS*
+
+1.  User requests to sort persons by field
+2.  AddressBook requests user for the field
+3.  User requests to sort persons by a specified field
+4.  AddressBook shows a list of persons by the fields in alphabetical order
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 3a. The given field is invalid
++
+Use case resumes at step 2.
+
+* 4a. The list is empty.
++
+Use case ends.
+
+[discrete]
+=== Use case: Sort person to user preference
+
+*MSS*
+
+1.  User requests to list persons
+2.  AddressBook shows a list of persons
+3.  User requests to list contacts in new order of preference
+4.  AddressBook shows a list of persons by user preference
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The list is empty.
++
+Use case ends.
+
+* 3a. The given index(es) is invalid.
++
+[none]
+** 3a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+
+* 3b. The given index(es) is repetitive.
++
+[none]
+** 3a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+
+[discrete]
+=== Use case: Find contacts by tags
+
+*MSS*
+
+1.  User requests to find all the tags
+2.  AddressBook shows a list of tags
+3.  User requests to list persons with specified tag
+4.  AddressBook shows a list of persons with the tag
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. There are no tags.
++
+Use case ends.
+
+* 3a. The given tag is invalid.
++
+[none]
+** 3a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+
+[discrete]
+=== Use case: Running Executable file
+
+*MSS*
+
+1.  User runs the executable file
+2.  AddressBook runs the software
++
+Use case ends.
+
 
 [appendix]
 == Non Functional Requirements
 
 .  Should work on any link:#mainstream-os[mainstream OS] as long as it has Java `1.8.0_60` or higher installed.
 .  Should be able to hold up to 1000 persons without a noticeable sluggishness in performance for typical usage.
 .  A user with above average typing speed for regular English text (i.e. not code, not system admin commands) should be able to accomplish most of the tasks faster using commands than using the mouse.
-
-{More to be added}
+.  Should come with automated unit tests and open source code.
+.  Should be able to log errors and backup data when errors occur.
+.  Should be lightweight, can respond back to user within 2 seconds.
+.  Should be user-friendly, usable even by computer novices.
```

Comment:
Not specific enough, you can say it like this "Non tech-savvy persons can use all commands correctly within 10 minutes of continuous interaction`. Make sure it's tangible.